### PR TITLE
Refactor finalize/reference processor

### LINKED ISSUE #145: [ospp][wip] Use the new weak reference processing API for the JikesRVM binding
this issue is tracking the weekly update on this project. also, [mmtk@zulipchat](https://mmtk.zulipchat.com/) has a sub-channel for this project discussion.

# Motivation
shortly,  porting the java-style jikesRVM to rust-lang is straightforward, but the reference processing is java-specific. 
say, we expected a more general reference processing if we want to move forward to another language or gc mechanism.

# Milestone
- [ ] Compile and familiar code base
- [ ] Refactor the weak ref handling mechanism
- [ ] Run DaCapo benchmark and analyze the perf result
- [ ] With previous perf result to optimize the current impl
- [ ] Verify the impl correctness on CI
- [ ] Write a doc

# reference
https://github.com/mmtk/mmtk-jikesrvm/issues/137
https://github.com/mmtk/mmtk-core/issues/694
https://summer-ospp.ac.cn/org/prodetail/235730136

### PR DESCRIPTION
**Pull request by fepicture**
_edit v1: sync latest status, i have finish the reference processor and integrate together._
_edit v2: update with binding PR link._

fix #145 
mmtk-jikesrvm PR https://github.com/mmtk/mmtk-jikesrvm/pull/150
jikesRVM PR https://github.com/mmtk/jikesrvm/pull/16

Modifications to the mmtk-jikesrvm repository encompass three key aspects:

1. I have removed some old bind APIs, marking them as unimplement, or directly removing them.
2. I implemented the process_weak_ref and forward_weak_ref methods. These will be invoked in the work bucket and act as an entry point for the original process in jikesrvm.
3. Lastly, the CI heap configuration has been updated.

### CODE REVIEW

**Comment by fepicture** [comment]
@qinsoon thanks for your commit to update jikesrvm version. 
the test suite heap usage has been tested on my local (64GiB RAM, i9-12900H), but I do not know why will fail :p

**Review by wks — CHANGES_REQUESTED** [review]
The code should function properly, although there may be some hidden problems that prevent some tests from passing.  But JikesRVM is still one of our officially-supported VMs.  We run binding tests for some PRs (will soon become all PRs) of mmtk-core, and JikesRVM is one of the VMs to run binding tests with.  So I think it is not appropriate to merge this unstable change and disable some tests now.

I suggest we merge your change, but keep your change disabled by default.  That is, the MMTK-JikesRVM will still use the reference processors in mmtk-core.  But with a compile-time flag, MMTk-JikesRVM will use JikesRVM's reference/finalizer processors, instead.

You can use conditional compilation tricks to enable/disable the code.

For Rust code, you can use Cargo features.  You can add a feature, such as `binding_side_ref_proc` in `Cargo.toml`.  Inside Rust source files, you can use the attribute `#[cfg(feature = "binding_side_ref_proc")]` to selectively enable types, functions, blocks, statements, etc.  You can also use `if cfg!(feature = "binding_side_ref_proc")` to guard the control flow.

For Java code, you can just write `if SOME_CONSTANT { ... }`, and JikesRVM's compiler is smart enough to eliminate dead code.  I'll discuss more in the review comments in your PR in the `jikesrvm` repo.

For test scripts such as `ci-test-assertions.sh`, instead of commenting out failing tests and changing heap sizes, I suggest copy those scripts, save to different files (such as `ci-test-assertions-weakref.sh`), and edit the content there, including disabling failing tests and changing heap sizes. (p.s. You should change both `-Xms` and `-Xmx` so that on each line the `-Xms` and `-Xmx` are set to the same value.)

And you should create a compile script to compile `jikesrvm` plus `mmtk-jikesrvm` using your added conditional compilation flags.

**Comment by fepicture** [comment]
hi @qinsoon， I have pushed a new commit to solve the comment

(basically, use conditional compile to keep the compatibility), 

can u activate the GitHub action workflow for this new commit?
 https://github.com/mmtk/mmtk-jikesrvm/actions/runs/6600465759

https://github.com/mmtk/mmtk-jikesrvm/actions/runs/6600465756

**Comment by qinsoon** [comment]
> hi @qinsoon， I have pushed a new commit to solve the comment
> 
> (basically, use conditional compile to keep the compatibility),
> 
> can u activate the GitHub action workflow for this new commit? https://github.com/mmtk/mmtk-jikesrvm/actions/runs/6600465759
> 
> https://github.com/mmtk/mmtk-jikesrvm/actions/runs/6600465756

Yeah. I just did. I thought the workflows would be allowed as they were approved for the previous commit. I will keep an eye on the PR.

**Comment by fepicture** [comment]
Thank you 🙇

File: mmtk/src/api.rs [hunk 1]
```diff
@@ -202,21 +201,16 @@ pub extern "C" fn modify_check(object: ObjectReference) {
...
-pub extern "C" fn add_weak_candidate(reff: ObjectReference, referent: ObjectReference) {
-    <JikesRVM as VMBinding>::VMReferenceGlue::set_referent(reff, referent);
-    memory_manager::add_weak_candidate(&SINGLETON, reff)
-}
-
-#[no_mangle]
-pub extern "C" fn add_soft_candidate(reff: ObjectReference, referent: ObjectReference) {
-    <JikesRVM as VMBinding>::VMReferenceGlue::set_referent(reff, referent);
-    memory_manager::add_soft_candidate(&SINGLETON, reff)
+pub extern "C" fn get_forwarded_object(object: ObjectReference) -> ObjectReference {
```
[the remark this answers is not in the document]
Comment (fepicture) [reply mmtk/src/api.rs:229]: Currently, I've come across a tricky issue in JikesRVM. We first need to define `xxfunctionRIP` as an address to facilitate linking, correct? However, this `xxfunctionRIP` is a class member, and I haven't yet discovered a straightforward method for conditional compilation for it. As a result, our current approach to conditional compilation isn't optimal:
- In the ideal case, based on the feature flag, we should compile certain files and exclude others, ensuring no additional build overhead.
- In our present implementation, even if the feature isn't enabled, some files still get compiled.

To summarize, functions like `get_forwarded_object` will be compiled regardless of whether we activate the feature.

[the remark this answers is not in the document]
Comment (wks) [reply mmtk/src/api.rs:229]: OK.  In this case, we can let such functions exist even when the vm-side forwarding feature is not enabled, as long as existing CI test cases still run properly.

**Review by wks — APPROVED** [review]
Although not all tests passed, the current state of the code used conditional compilation and other techniques to keep the CI tests passing.  In this sense, it looks good to me.

We should keep working on this feature to make it stable so that we can eventually use VM-side weak ref and finalization processing by default.

**Comment by wks** [comment]
A recent upstream change moved the method `is_emergency_collection` from `Plan` to `GlobalState`, but `GlobalState` is a private structure and inaccessible from the VM binding.  That prevents this PR from working.

The built-in reference processors and finalization processors can access the `MMTK::state` field because they are parts of the `mmtk` crate.  We should make change to `mmtk-core` and expose the `is_emergency_collection` function to the binding.

**Comment by fepicture** [comment]
> A recent upstream change (https://github.com/mmtk/mmtk-core/commit/57af17fbfd94ff0df2cd3b1e504abe299ce4f0ab) moved the method `is_emergency_collection` from `Plan` to `GlobalState`, but `GlobalState` is a private structure and inaccessible from the VM binding.  That prevents this PR from working.
> 
> The built-in reference processors and finalization processors can access the `MMTK::state` field because they are parts of the `mmtk` crate.  We should make change to `mmtk-core` and expose the `is_emergency_collection` function to the binding.

thanks I will fix later.

**Comment by fepicture** [comment]
hi, @qinsoon 

I have pushed a new commit to update with mmtk-core PR(expose the API for check emergency collection )

Plz activate the GitHub action workflow.

**Comment by wks** [comment]
The "Dacapo Tests / test" failed with a known bug that existed before this PR was introduced.  I'll rerun the test.
